unipro.c

```diff
@@ -0,0 +1,571 @@
+};
+
+static struct ufs_unipro_attrs_info uic_attrs_group[MAX_UNIPRO_IDN] = {
+	{
+		"MIPI M-PHY", mipi_mphy_attrs,
+		sizeof(mipi_mphy_attrs) / sizeof(struct ufs_uic_attr_fields)
+	},
+	{
+		"PHY-Adapter", phy_adapter_attrs,
+		sizeof(phy_adapter_attrs) / sizeof(struct ufs_uic_attr_fields)
+	},
+	{
+		"DME Attributes for QoS", dme_qos_attrs,
+		sizeof(dme_qos_attrs) / sizeof(struct ufs_uic_attr_fields)
+	},
+};
+
+static struct uic_cmd_result_code resultcode[] = {
+	{0, "SUCCESS"},
+	{1, "INVALID_MIB_ATTRIBUTE"},
+	{2, "INVALID_MIB_ATTRIBUTE_VALUE"},
+	{3, "READ_ONLY_MIB_ATTRIBUTE"},
+	{4, "WRITE_ONLY_MIB_ATTRIBUTE"},
+	{5, "BAD_INDEX"},
+	{6, "LOCKED_MIB_ATTRIBUTE"},
+	{7, "BAD_TEST_FEATURE_INDEX"},
+	{8, "PEER_COMMUNICATION_FAILURE"},
+	{9, "BUSY"},
+	{10, "DME_FAILURE"},
+};
+
+static int ufshcd_dme_get_attr(int fd, __u32 attr_sel, __u8 peer)
+{
+	struct ufs_bsg_request bsg_req = { 0 };
+	struct ufs_bsg_reply bsg_rsp = { 0 };
+	struct uic_command *uic_cmd =
+		(struct uic_command *)&bsg_req.upiu_req.uc;
+	struct uic_command uic_rsq = { 0 };
+
+	int rt = OK;
+	__u8 res_code;
+
+	uic_cmd->command = peer ? UIC_CMD_DME_PEER_GET : UIC_CMD_DME_GET;
+	uic_cmd->argument1 = attr_sel;
+	bsg_req.msgcode = UPIU_TRANSACTION_UIC_CMD;
+
+	rt = send_bsg_sg_io(fd, &bsg_req, &bsg_rsp, 0, 0, 0);
+	if (rt) {
+		print_error("%s: bsg request failed.\n", __func__);
+		rt = ERROR;
+		goto out;
+	}
+
+	memcpy(&uic_rsq, &bsg_rsp.upiu_rsp.uc, UIC_CMD_SIZE);
+	res_code = uic_rsq.argument2 & MASK_UIC_COMMAND_RESULT;
+
+	if (res_code) {
+		__u8 max_code =
+		sizeof(resultcode) /
+			sizeof(struct uic_cmd_result_code);
+
+		if (res_code < (max_code - 1)) {
+			print_error("%s: attr-id 0x%x %s.\n",
+				    __func__,
+				    UIC_GET_ATTR_ID(attr_sel),
+				    resultcode[res_code].def);
+		} else {
+			print_error("%s: ID 0x%x, unknown error code %d.\n",
+				    __func__, UIC_GET_ATTR_ID(attr_sel),
+				    res_code);
+		}
+
+		rt = ERROR;
+	} else {
+		rt = uic_rsq.argument3;
+	}
+
+out:
+	return rt;
+}
+
+static int ufshcd_dme_set_attr(int fd, __u32 attr_sel, __u8 attr_set,
+			       __u32 mib_val, __u8 peer)
+{
+	struct ufs_bsg_request bsg_req = { 0 };
+	struct ufs_bsg_reply bsg_rsp = { 0 };
+	struct uic_command *uic_cmd =
+		(struct uic_command *)&bsg_req.upiu_req.uc;
+	struct uic_command uic_rsq = { 0 };
+
+	int rt = OK;
+	__u8 res_code;
+
+	uic_cmd->command = peer ? UIC_CMD_DME_PEER_SET : UIC_CMD_DME_SET;
+	uic_cmd->argument1 = attr_sel;
+	uic_cmd->argument2 = UIC_ARG_ATTR_TYPE(attr_set);
+	uic_cmd->argument3 = mib_val;
+
+	bsg_req.msgcode = UPIU_TRANSACTION_UIC_CMD;
+
+	rt = send_bsg_sg_io(fd, &bsg_req, &bsg_rsp, 0, 0, 0);
+	if (rt) {
+		print_error("%s: bsg request failed.\n", __func__);
+		rt = ERROR;
+		goto out;
+	}
+
+	memcpy(&uic_rsq, &bsg_rsp.upiu_rsp.uc, UIC_CMD_SIZE);
+	res_code = uic_rsq.argument2 & MASK_UIC_COMMAND_RESULT;
+
+	if (res_code) {
+		__u8 max_code = sizeof(resultcode) /
+			sizeof(struct uic_cmd_result_code);
+
```

Comment:
Maybe you can use ARRAY_SIZE define

Reply:
hi, Arthur
I saw this macro you defined in ufs_cmds.c.
#define ARRAY_SIZE(x) (sizeof(x) / sizeof((x)[0]))

If I change this, I should move this macro to .h. then I can change, or I define it in unipro.c again.
that will be redundant.  I want firstly keep my current code, let me fix it later if you are ok. 

thanks for your suggestion, let me see how other reviewers think about.